pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,363 @@
+	workload *v1alpha1.DownsyncObjectReferences,
+	namespaceScopeDownsyncObjectsMap map[schema.GroupVersionResource]*v1alpha1.NamespaceScopeDownsyncObjects,
+	nsObjectsLocationInSlice map[string]int) {
+	if nsdObjects, found := namespaceScopeDownsyncObjectsMap[gvr]; found {
+		// GVR mapping is found, check NS mapping
+		gvrAndNSKey := util.KeyFromGVRandNS(gvr, key.NamespacedName.Namespace) // key for ns mapping inside
+
+		if nsIdx, found := nsObjectsLocationInSlice[gvrAndNSKey]; found {
+			// NS mapping is found, append name
+			nsdObjects.ObjectsByNamespace[nsIdx].Names = append(nsdObjects.ObjectsByNamespace[nsIdx].Names,
+				key.NamespacedName.Name)
+			return
+		}
+
+		// namespace mapping is not found, create new (ns, names) entry for object
+		nsdObjects.ObjectsByNamespace = append(nsdObjects.ObjectsByNamespace, v1alpha1.NamespaceAndNames{
+			Namespace: key.NamespacedName.Namespace,
+			Names:     []string{key.NamespacedName.Name},
+		})
+
+		// update index mapping
+		nsObjectsLocationInSlice[gvrAndNSKey] = len(nsdObjects.ObjectsByNamespace) - 1
+		return
+	}
+
+	// GVR mapping isn't found, create new entry for this obj (and NS map)
+	// add namespaceScopeDownsyncObjectsMap to the workload
+	workload.NamespaceScope = append(workload.NamespaceScope, v1alpha1.NamespaceScopeDownsyncObjects{
+		GroupVersionResource: metav1.GroupVersionResource{
+			Group:    gvr.Group,
+			Version:  gvr.Version,
+			Resource: gvr.Resource,
+		},
+		ObjectsByNamespace: []v1alpha1.NamespaceAndNames{
+			{
+				Namespace: key.NamespacedName.Namespace,
+				Names:     []string{key.NamespacedName.Name},
+			},
+		},
+	})
+
+	// retain a pointer to the added namespaceScopeDownsyncObjectsMap for efficiency
+	namespaceScopeDownsyncObjectsMap[gvr] = &workload.NamespaceScope[len(workload.NamespaceScope)-1]
+	// update ns mapping
+	nsObjectsLocationInSlice[util.KeyFromGVRandNS(gvr, key.NamespacedName.Namespace)] = 0 // first entry
+}
+
+func (resolution *placementResolution) matchesPlacementDecisionSpec(placementDecisionSpec *v1alpha1.PlacementDecisionSpec,
+	gvkGvrMapper util.GvkGvrMapper) bool {
+	resolution.RLock()
+	defer resolution.RUnlock()
+
+	// check destinations
+	if !destinationsMatch(resolution.destinations, placementDecisionSpec.Destinations) {
+		return false
+	}
+
+	// check workload
+	return workloadMatchesPlacementDecisionSpec(&placementDecisionSpec.Workload, resolution.objectIdentifierToKeyMap, gvkGvrMapper)
+}
+
+// destinationsMatch returns true if the destinations in the resolution
+// match the destinations in the placement decision spec.
+// The function performs one pass over both slices, sequentially.
+func destinationsMatch(destinations []string, placementDecisionDestinations []v1alpha1.Destination) bool {
+	if len(destinations) != len(placementDecisionDestinations) {
+		return false
+	}
+
+	// calculate match by counting matching objects then comparing to length.
+	// this works since entries are unique.
+	matches := map[string]interface{}{}
```

Comment:
`PlacementDecision` objects will be easier for users to consume if the lists are kept in alphabetical order.

Reply:
I dont understand the correlation between this comment and marked code

Reply:
It is somewhat distant. It was triggered by thinking about more efficiently comparing slices. But that is moot, now the comparison is between a map and a slice.

Even so, the thought about consistently ordering the slice contents has merit on its own. When users go looking for changes, they will be easier to find if there is consistent ordering in the slices.

Reply:
I opened #1716 for the ordering.